### PR TITLE
re-push "add volume kubelet_volume_stats_health_abnormal to kubelet #105585"

### PR DESCRIPTION
**Pull request by fengzixu**
#### What type of PR is this?

/kind feature
/kind bug

#### What this PR does / why we need it:

- add `kubelet_volume_stats_health_abnormal` metrics to kubelet based on this [kep](https://github.com/kubernetes/enhancements/pull/2900)

- Fix panic bug when volumeHealthStatus is nil on https://github.com/kubernetes/kubernetes/issues/108715#issuecomment-1069085550

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/enhancements/pull/2900

#### Does this PR introduce a user-facing change?

```release-note
add one metrics(`kubelet_volume_stats_health_abnormal`) of volume health state to kubelet
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/pull/2900

```

### CODE REVIEW

**Comment by fengzixu** [comment]
oh. I see. the previous PR is reverted. I shouldn't rebase the current master branch

File: pkg/kubelet/metrics/collectors/volume_stats.go [hunk 1]
```diff
@@ -120,6 +120,12 @@ func (collector *volumeStatsCollector) CollectWithStability(ch chan<- metrics.Me
 			addGauge(volumeStatsInodesDesc, pvcRef, float64(*volumeStat.Inodes))
 			addGauge(volumeStatsInodesFreeDesc, pvcRef, float64(*volumeStat.InodesFree))
 			addGauge(volumeStatsInodesUsedDesc, pvcRef, float64(*volumeStat.InodesUsed))
+<<<<<<< Updated upstream
+=======
+			if volumeStat.VolumeHealthStats != nil {
+				addGauge(volumeStatsHealthAbnormalDesc, pvcRef, convertBoolToFloat64(volumeStat.VolumeHealthStats.Abnormal))
+			}
+>>>>>>> Stashed changes
```
Comment (xing-yang) [inline pkg/kubelet/metrics/collectors/volume_stats.go]: Can you squash the "[resolve conflict](https://github.com/kubernetes/kubernetes/pull/108758/commits/afdbd27abfef8a10cc4ba389dbc2fd68ba70535f)" commit into this one?

Comment (fengzixu) [reply pkg/kubelet/metrics/collectors/volume_stats.go]: sure

File: pkg/kubelet/server/stats/volume_stat_calculator_test.go [hunk 2]
```diff
@@ -263,6 +267,13 @@ func expectedFSStats() kubestats.FsStats {
 	}
 }
 
+func expectedVolumeHealthStats() *kubestats.VolumeHealthStats {
+	metric := expectedMetrics()
```
Comment (logicalhan) [inline pkg/kubelet/server/stats/volume_stat_calculator_test.go:271]: shouldn't you have the nil check here too?

Comment (fengzixu) [reply pkg/kubelet/server/stats/volume_stat_calculator_test.go:271]: no. expectedMetrics is a function that is only used in test: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/stats/volume_stat_calculator_test.go#L235

Comment (xing-yang) [reply pkg/kubelet/server/stats/volume_stat_calculator_test.go:271]: Even if it is a test, I think we should still check nil to be safe.

File: pkg/kubelet/server/stats/volume_stat_calculator_test.go [hunk 3]
```diff
@@ -263,6 +267,13 @@ func expectedFSStats() kubestats.FsStats {
...
+	return &kubestats.VolumeHealthStats{
+		Abnormal: *metric.Abnormal,
+	}
```
Comment (xing-yang) [inline pkg/kubelet/server/stats/volume_stat_calculator_test.go:274]: We should check both metric and metric.Abnormal are not nil.

Also check if Abnormal or Message is nil before dereferencing them:
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.4/pkg/kubelet/server/stats/volume_stat_calculator_test.go#L243-L244

**Review by mrunalp — APPROVED** [review]

**Pull request merged**